go/state/state_db_test.go

```diff
@@ -4143,6 +4144,135 @@ func TestSlotIdOrder(t *testing.T) {
+	}
+
+	db.RevertToSnapshot(s)
+
+	if got, want := db.GetTransientState(address1, key1), want; got != want {
+		t.Errorf("unexpected value, wanted %v, got %v", want, got)
+	}
+}
+
+func TestStateDB_TransientStorage_IsClearedAfterCallingEndTransaction(t *testing.T) {
+	ctrl := gomock.NewController(t)
+	mock := NewMockState(ctrl)
+	db := createStateDBWith(mock, defaultStoredDataCacheSize, true)
+
+	db.SetTransientState(address1, key1, val1)
+	// Check that value was correctly set
+	if got, want := db.GetTransientState(address1, key1), val1; got != want {
+		t.Errorf("unexpected value, wanted %v, got %v", want, got)
+	}
+
+	db.EndTransaction()
+	if got, want := db.GetTransientState(address1, key1), valEmpty; got != want {
+		t.Errorf("unexpected value, wanted %v, got %v", want, got)
+	}
+
+	if len(db.undo) > 0 {
+		t.Errorf("unexpected undo len, wanted: 0, got: %v", len(db.undo))
+	}
+}
+
+func TestStateDB_SetTransientState_NewestValueCanBeObtained(t *testing.T) {
+	ctrl := gomock.NewController(t)
+	mock := NewMockState(ctrl)
+	db := CreateStateDBUsing(mock)
+
+	db.SetTransientState(address1, key1, val1)
+	if got, want := db.GetTransientState(address1, key1), val1; got != want {
+		t.Errorf("unexpected value, wanted %v, got %v", want, got)
+	}
+
+	db.SetTransientState(address1, key1, val2)
+	if got, want := db.GetTransientState(address1, key1), val2; got != want {
+		t.Errorf("unexpected value, wanted %v, got %v", want, got)
+	}
+}
+
+func TestStateDB_SetTransientState_ValueIsNotSetIfSame(t *testing.T) {
+	ctrl := gomock.NewController(t)
+	mock := NewMockState(ctrl)
+	db := CreateStateDBUsing(mock)
+
+	db.SetTransientState(address1, key1, val1)
+	if got, want := db.GetTransientState(address1, key1), val1; got != want {
+		t.Errorf("unexpected value, wanted %v, got %v", want, got)
+	}
+
+	db.SetTransientState(address1, key1, val2)
+	if got, want := db.GetTransientState(address1, key1), val2; got != want {
+		t.Errorf("unexpected value, wanted %v, got %v", want, got)
+	}
+}
+
+func TestStateDB_SetTransientState_SettingSameValueReturnsEarly(t *testing.T) {
+	ctrl := gomock.NewController(t)
+	mock := NewMockState(ctrl)
+	db := createStateDBWith(mock, defaultStoredDataCacheSize, true)
+
+	db.SetTransientState(address1, key1, val1)
+	if got, want := db.GetTransientState(address1, key1), val1; got != want {
+		t.Errorf("unexpected value, wanted %v, got %v", want, got)
+	}
+
+	if got, want := len(db.undo), 1; got != want {
+		t.Errorf("unexpected undo len, wanted: 1, got: %v", len(db.undo))
+	}
+
+	undo := db.undo[0]
+
+	db.SetTransientState(address1, key1, val1)
+	if got, want := db.GetTransientState(address1, key1), val1; got != want {
+		t.Errorf("unexpected value, wanted %v, got %v", want, got)
+	}
+
+	if got, want := len(db.undo), 1; got != want {
+		t.Errorf("unexpected undo len, wanted: 1, got: %v", len(db.undo))
+	}
+
+	if got, want := reflect.ValueOf(db.undo[0]), reflect.ValueOf(undo); got != want {
```

Comment:
perhaps not necessary to test this - it is too unlikely

Reply:
You are right. This can probably never be false.